torch/csrc/autograd/python_function.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <Python.h>
+#include <vector>
+#include <utility>
+
+#include "torch/csrc/autograd/function.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/utils/object_ptr.h"
+
+// (class, gpu id, sizes)
+using output_info_type = std::tuple<PyObject *, int, std::vector<long>>;
+// (tensor, version when saved, version counter)
+// or
+// (None, 0, nullptr)
+using saved_var_info_type = std::tuple<THPObjectPtr, int, std::unique_ptr<torch::autograd::VariableVersion>>;
+
+namespace torch { namespace autograd {
+
+struct PyFunction : public Function {
+  PyFunction(PyObject* obj) : obj(obj) {}
+
+  virtual variable_list apply(const variable_list& inputs) override;
+  virtual void releaseVariables() override;
+
+  PyObject* obj;
+};
+
+}} // namespace torch::autograd
+
+struct THPFunction {
+    PyObject_HEAD
+
+    PyObject *needs_input_grad;
+    PyObject *backward_hooks;
+    THPObjectPtr *output_backward_hooks;
+
+    PyObject *to_save;
+    PyObject *shared_pairs;
+    PyObject *non_differentiable;
+    PyObject *dirty_tensors;
+
+    std::vector<output_info_type> *output_info;
+    std::vector<saved_var_info_type> *saved_variables;
+    int num_inputs;
+    char has_freed_buffers;
+
+    torch::autograd::PyFunction impl;
```

Comment:
Maybe it would be more consistent to call it `cdata`. But it's ok as is I think.